modules/azure-auth-header/azure-auth-header.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2020 One Identity
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "azure-auth-header.h"
+#include "driver.h"
+#include "modules/http/http-signals.h"
+#include "messages.h"
+#include "compat/openssl_support.h"
+
+#define AZURE_AUTH_HEADER_PLUGIN "azure-auth-header"
+
+#define MAX_DATE_LEN 64
+
+struct _AzureAuthHeaderPlugin
+{
+  LogDriverPlugin super;
+
+  gsize   secret_len;
+  guchar *secret;
+
+  gchar   *workspace_id;
+  gchar   *method;
+  gchar   *path;
+  gchar   *content_type;
+};
+
+static gsize
+_get_rfc1123date(gchar *buf, gsize buf_len)
+{
+  time_t now = time(NULL);
+  struct tm gmt;
+
+  gmtime_r(&now, &gmt);
+
+  gchar format[] = "%a, %d %b %Y %H:%M:%S GMT";
+  gsize len = strftime(buf, buf_len, format, &gmt);
+
+  return len;
+}
+
+static void
+_get_rfc1123data_or_assert(gchar *date)
+{
+  gsize date_len = 0;
+  date_len = _get_rfc1123date(date, MAX_DATE_LEN);
+  g_assert(date_len);
+}
+
+static GString *
+_azure_auth_header_get_str_to_hash(AzureAuthHeaderPlugin *self, glong content_len, const gchar *date)
+{
+  gchar *fmt = "%s\n%ld\n%s\nx-ms-date:%s\n%s";
+  GString *str = g_string_new(NULL);
+
+  g_string_append_printf(str, fmt, self->method, content_len, self->content_type, date, self->path);
+
+  return str;
+}
+
+static guint
+_azure_auth_header_get_digest(AzureAuthHeaderPlugin *self, GString *input, guchar *digest)
+{
+  guint md_len = 0;
+
+  if (!HMAC(EVP_sha256(),
+            self->secret, self->secret_len,
+            (const guchar *)input->str, input->len,
+            digest, &md_len))
+    {
+      msg_error("Failed to generate Azure Auth Header HMAC",
+                evt_tag_str("str", input->str),
+                evt_tag_int("len", input->len));
+    }
+
+  return md_len;
```

Comment:
In case `HMAC()` fails, I think it would be better to return early with an explicit 0, because it is not specified by the documentation whether `md_len` is undefined or untouched when the operation fails.

Reply:
Well... it's not a secret that `HMAC` is a helper function around the `HMAX_CTX` (new, init, update, final, free). The `HMAC_Final` needs to check whether the mdlen is `NULL` (just check the doc) ... which should mean that it is set only after everything went fine. So even it is `not specified` it will work as expected.
But I can early return...